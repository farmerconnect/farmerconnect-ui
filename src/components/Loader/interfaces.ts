import { ReactNode } from 'react';
import CustomStyles from '../CustomStyles';

export interface ILoaderIcon {
  size: number;
  color: string;
}

interface ILoaderButton {
  color: string;
}

export interface ILoader {
  icon: ILoaderIcon;
  button: ILoaderButton;
}

interface ILoaderOption {
  message: string;
  duration?: number;
}

interface ILoaderError {
  error: boolean;
  message: ReactNode;
}

export interface ILoaderProps extends CustomStyles {
  show?: boolean;
  options?: ILoaderOption[];
  iconSize?: number;
  iconColor?: string;
  buttonText?: string | ReactNode;
  errorOptions?: ILoaderError;
  onButtonClick?: () => void;
}
