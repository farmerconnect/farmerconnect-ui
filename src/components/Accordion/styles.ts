import { IAccordionProps } from './interfaces';
import styled from 'styled-components';
import { farmerConnectTheme } from '../Theme';

export const Container = styled.div<IAccordionProps>`
  box-sizing: border-box;
  border-radius: 0.75rem;
  border: 1px solid ${farmerConnectTheme.colors.fc_black_10};

  header {
    box-sizing: border-box;
    padding: 0.5rem 0.625rem;
    display: flex;
    align-items: center;
    cursor: default;
    font-size: 0.875rem;
    font-weight: 500;

    > svg {
      color: ${farmerConnectTheme.colors.fc_black_70};
      margin: 0 1rem 0 0.25rem;
      transition: transform 0.1s ease-out;
      cursor: pointer;
      transform: ${({ isOpen }: IAccordionProps) => (isOpen ? 'rotate(0deg)' : 'rotate(-90deg)')};
    }
  }

  main {
    box-sizing: border-box;
    padding: 1rem 2.25rem 3.5rem;
  }

  & + & {
    border-top: none;
  }
`;
