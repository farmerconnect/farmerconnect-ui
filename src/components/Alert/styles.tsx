import styled from 'styled-components';
import colors from '../../styles/colors';
import { IAlertProps } from './interfaces';

const colorVariant = {
  warning: {
    background: colors.fc_light_yellow,
    foreground: colors.fc_dark_yellow,
  },
  info: {
    background: colors.fc_light_blue,
    foreground: colors.fc_dark_blue,
  },
  danger: {
    background: colors.fc_light_red,
    foreground: colors.fc_hover_red,
  },
} as const;

export const Container = styled.div<IAlertProps>`
  margin: 0;
  position: relative;
  display: inline-block;
  border-radius: 0.75rem;
  padding: 0.5rem 1rem 0.5rem 2.5rem;
  background-color: ${({ variant }) =>
    variant && colorVariant[variant] ? colorVariant[variant].background : colorVariant.warning.background};
  > svg.alert-icon {
    max-height: 1rem;
    width: 1rem;
    color: ${({ variant }) =>
      variant && colorVariant[variant] ? colorVariant[variant].foreground : colorVariant.warning.foreground};
    position: absolute;
    left: 1rem;
    top: 50%;
    transform: translateY(-50%);
  }
`;
