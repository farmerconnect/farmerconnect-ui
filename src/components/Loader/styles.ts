import styled, { css, keyframes } from 'styled-components';
import { ILoader, ILoaderProps } from './interfaces';
import StyledButton from '../Button';

export const loaderDefaultStyles: ILoader = {
  icon: {
    color: '#00E394',
    size: 158,
  },
  button: {
    color: '#6D6D6D',
  },
};

const originalSize = 231;

const scaleTo = (number: number, to: number = loaderDefaultStyles.icon.size) =>
  (number * ((to * 100) / originalSize)) / 100;

const loadingAnimation = keyframes`
  0% { opacity: 1 }
  100% { opacity: 0 }
`;

export const Button = styled(StyledButton)`
  color: ${({ theme, color }) => color || theme?.loader?.button?.color || loaderDefaultStyles.button.color};
  font-size: 14px;
  line-height: 18px;
  margin-top: 80px;
  text-decoration: underline;

  &,
  &:hover {
    background-color: transparent;
  }
`;

export const Message = styled.p`
  font-size: 18px;
  line-height: 24px;
  margin: 20px 0 0;
  text-align: center;

  &.error {
    max-width: 392px;
  }

  a {
    color: black;
  }
`;

export const Content = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-top: 0;
  opacity: 1;
  transition: opacity 0.5s, margin-top 0s 0s;
`;

export const Container = styled.div`
  align-items: center;
  background-color: rgba(255, 255, 255, 0.9);
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: center;
  left: 0;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 999;

  ${({ show }: ILoaderProps) =>
    !show &&
    css`
      transition: visibility 0s 0.5s;
      visibility: hidden;

      ${Content} {
        opacity: 0;
        margin-top: -100vh;
        transition: opacity 0.5s, margin-top 0s 0.5s;
      }
    `}
`;

export const Spinner = styled.div<ILoaderProps>`
  display: inline-block;
  overflow: hidden;
  background: none;
  flex-shrink: 0;

  ${({ iconSize }: ILoaderProps) => css`
    height: ${scaleTo(231, iconSize)}px;
    width: ${scaleTo(231, iconSize)}px;
  `}
`;

export const SpinnerItemList = styled.div`
  width: 100%;
  height: 100%;
  position: relative;
  transform: translateZ(0) scale(1);
  backface-visibility: hidden;
  transform-origin: 0 0;
`;

export const SpinnerItem = styled.div<ILoaderProps>`
  box-sizing: content-box;
  position: absolute;
  animation: ${loadingAnimation} linear 1.0204081632653061s infinite;
  background: ${({ theme, iconColor }) => iconColor || theme?.loader?.color || loaderDefaultStyles.icon.color};

  ${({ iconSize }: ILoaderProps) => css`
    left: ${scaleTo(103.95, iconSize)}px;
    top: ${scaleTo(36.96, iconSize)}px;
    width: ${scaleTo(23.1, iconSize)}px;
    height: ${scaleTo(23.1, iconSize)}px;
    border-radius: ${scaleTo(6.93, iconSize)}px / ${scaleTo(6.93, iconSize)}px;
    transform-origin: ${scaleTo(11.55, iconSize)}px ${scaleTo(78.54, iconSize)}px;
  `}

  &:nth-child(1) {
    transform: rotate(0deg);
    animation-delay: -0.9070294784580498s;
  }

  &:nth-child(2) {
    transform: rotate(40deg);
    animation-delay: -0.7936507936507936s;
  }

  &:nth-child(3) {
    transform: rotate(80deg);
    animation-delay: -0.6802721088435374s;
  }

  &:nth-child(4) {
    transform: rotate(120deg);
    animation-delay: -0.5668934240362812s;
  }

  &:nth-child(5) {
    transform: rotate(160deg);
    animation-delay: -0.4535147392290249s;
  }

  &:nth-child(6) {
    transform: rotate(200deg);
    animation-delay: -0.3401360544217687s;
  }

  &:nth-child(7) {
    transform: rotate(240deg);
    animation-delay: -0.22675736961451246s;
  }

  &:nth-child(8) {
    transform: rotate(280deg);
    animation-delay: -0.11337868480725623s;
  }

  &:nth-child(9) {
    transform: rotate(320deg);
    animation-delay: 0s;
  }
`;
