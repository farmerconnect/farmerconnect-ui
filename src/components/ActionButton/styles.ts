import styled from 'styled-components';
import {Button as CustomButton} from '../CustomButton/styles';

export const IconWrapper = styled(CustomButton)`
  margin: 0px;
  padding: 0px;
  height: 32px;
  width: 32px;
  display: flex;
  border-radius: 8px
`
