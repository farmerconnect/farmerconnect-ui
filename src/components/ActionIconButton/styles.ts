import styled, { css } from 'styled-components';
import {Button as CustomButton} from '../CustomButton/styles';

const iconOnlyStyle = css`
  margin: 0px;
  padding: 0px;
  height: 32px;
  width: 32px;
  display: flex;
  border-radius: 8px
`

export const ActionIconButton = styled(CustomButton)`
  ${iconOnlyStyle}
`
