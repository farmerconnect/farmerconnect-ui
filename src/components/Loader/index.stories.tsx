import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import Loader from './';
import { loaderDefaultStyles } from './styles';

export default {
  title: 'Loader',
  component: Loader,
  argTypes: {
    show: { control: 'boolean' },
    options: { control: 'object' },
    iconSize: { control: 'number' },
    iconColor: { control: 'color' },
  },
} as Meta;

const defaultArgs = {
  show: true,
  options: [
    {
      duration: 5,
      message: 'We are working on your request, please bare with us for a moment...',
    },
    {
      duration: 5,
      message: '...just a little bit longer',
    },
    {
      duration: 5,
      message: '...almost there',
    },
  ],
  iconColor: loaderDefaultStyles.icon.color,
  buttonText: 'Cancel request',
};

const Template: Story = (args) => <Loader {...args} />;

export const Default = Template.bind({});
Default.args = { ...defaultArgs };

export const CustomIconSize = Template.bind({});
CustomIconSize.args = { ...defaultArgs, iconSize: 50 };
