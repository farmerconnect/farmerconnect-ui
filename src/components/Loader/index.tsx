import React, { Fragment, useCallback, useEffect, useMemo, useState } from 'react';
import { ILoaderProps } from './interfaces';
import Caution from '../Icons/Caution';
import { Button, Content, Message, Spinner, Container, SpinnerItem, SpinnerItemList } from './styles';

const Loader = ({
  show,
  options,
  iconSize,
  iconColor,
  buttonText,
  errorOptions,
  onButtonClick,
  ...props
}: ILoaderProps) => {
  const [currentIndex, setCurrentIndex] = useState<number>(0);
  const option = useMemo(() => options && options[currentIndex], [options, currentIndex]);

  useEffect(() => {
    if (currentIndex !== 0 && !show) setCurrentIndex(0);
  }, [currentIndex, show]);

  useEffect(() => {
    if (!options?.length || !show) return;

    let timeout: ReturnType<typeof setTimeout>;
    const hasNextElement = currentIndex < options.length - 1;

    if (option && option.duration && hasNextElement) {
      timeout = setTimeout(() => {
        setCurrentIndex(currentIndex + 1);
      }, option.duration * 1000);
    }

    return () => timeout && clearTimeout(timeout);
  }, [currentIndex, option, options, show]);

  const renderSpinner = useCallback(
    () => (
      <Spinner iconSize={iconSize}>
        <SpinnerItemList>
          {Array(9)
            .fill('')
            .map((_, index: number) => (
              <SpinnerItem key={index} iconSize={iconSize} iconColor={iconColor} />
            ))}
        </SpinnerItemList>
      </Spinner>
    ),
    [iconColor, iconSize]
  );

  const renderMessage = useCallback(() => (option ? <Message>{option?.message}</Message> : null), [option]);

  const renderError = useCallback(() => {
    return (
      <Fragment>
        <Caution />
        <Message className="error">{errorOptions?.message}</Message>
      </Fragment>
    );
  }, [errorOptions]);

  return (
    <Container show={show} {...props}>
      <Content>
        {errorOptions?.error ? renderError() : renderSpinner()}
        {errorOptions?.error ? '' : renderMessage()}
        {onButtonClick && buttonText && <Button onClick={onButtonClick}>{buttonText}</Button>}
      </Content>
    </Container>
  );
};

Loader.defaultProps = {
  show: false,
};

export default Loader;
