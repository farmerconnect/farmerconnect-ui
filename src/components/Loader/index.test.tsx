import React from 'react';
import { shallow } from 'enzyme';
import Loader from './';
import { Button } from './styles';

describe('Loader component', () => {
  const defaultProps = {
    show: false,
    options: [
      {
        duration: 2,
        message: 'We are working on your request, please bare with us for a moment...',
      },
      {
        duration: 2,
        message: '...just a little bit longer',
      },
      {
        duration: 2,
        message: '...almost there',
      },
    ],
    buttonText: 'Cancel request',
    onButtonClick: jest.fn(),
  };

  it('should match the snapshot - hidden', () => {
    const wrapper = shallow(<Loader {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot - visible', () => {
    const wrapper = shallow(<Loader {...defaultProps} show />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the spy function', () => {
    const wrapper = shallow(<Loader {...defaultProps} show />);
    expect(wrapper.find(Button).text()).toBe(defaultProps.buttonText);

    wrapper.find(Button).simulate('click');
    expect(defaultProps.onButtonClick).toHaveBeenCalled();
  });
});
