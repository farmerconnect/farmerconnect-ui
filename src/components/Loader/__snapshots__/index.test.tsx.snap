// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader component should match the snapshot - hidden 1`] = `
<styled.div
  show={false}
>
  <styled.div>
    <styled.div>
      <styled.div>
        <styled.div
          key="0"
        />
        <styled.div
          key="1"
        />
        <styled.div
          key="2"
        />
        <styled.div
          key="3"
        />
        <styled.div
          key="4"
        />
        <styled.div
          key="5"
        />
        <styled.div
          key="6"
        />
        <styled.div
          key="7"
        />
        <styled.div
          key="8"
        />
      </styled.div>
    </styled.div>
    <styled.p>
      We are working on your request, please bare with us for a moment...
    </styled.p>
    <Styled(Button)
      onClick={[MockFunction]}
    >
      Cancel request
    </Styled(Button)>
  </styled.div>
</styled.div>
`;

exports[`Loader component should match the snapshot - visible 1`] = `
<styled.div
  show={true}
>
  <styled.div>
    <styled.div>
      <styled.div>
        <styled.div
          key="0"
        />
        <styled.div
          key="1"
        />
        <styled.div
          key="2"
        />
        <styled.div
          key="3"
        />
        <styled.div
          key="4"
        />
        <styled.div
          key="5"
        />
        <styled.div
          key="6"
        />
        <styled.div
          key="7"
        />
        <styled.div
          key="8"
        />
      </styled.div>
    </styled.div>
    <styled.p>
      We are working on your request, please bare with us for a moment...
    </styled.p>
    <Styled(Button)
      onClick={[MockFunction]}
    >
      Cancel request
    </Styled(Button)>
  </styled.div>
</styled.div>
`;
